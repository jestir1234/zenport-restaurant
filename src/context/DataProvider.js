import React, { Component } from "react";
import data from "../data.json";

export const DataContext = React.createContext();

const parseRestaurants = () => {
  let restaurants = {};
  let uniqRestaurants = Array.from(
    new Set(data[0].dishes.map(dish => dish.restaurant))
  );
  uniqRestaurants.forEach(name => {
    restaurants[name] = { availableMeals: "", menu: [] };
  });

  data[0].dishes.forEach(dish => {
    let currentMenu = restaurants[dish.restaurant].menu;
    currentMenu.push(dish.name);
    restaurants[dish.restaurant] = {
      availableMeals: dish.availableMeals,
      menu: currentMenu
    };
  });
  return restaurants;
};

const parseMealTypes = restaurants => {
  let mealTypes = {};
  Object.keys(restaurants).forEach(restaurant => {
    let restaurantMealTypes = restaurants[restaurant]["availableMeals"];
    restaurantMealTypes.forEach(type => {
      if (mealTypes[type]) {
        mealTypes[type]["restaurants"][restaurant] = restaurants[restaurant];
      } else {
        mealTypes[type] = { restaurants: {} };
      }
    });
  });
  return mealTypes;
};

const defaultStepOptions = () => ({
  step1: {
    selectedMealType: null,
    selectedPartyCount: { value: 1, label: 1 }
  },
  step2: {
    selectedRestaurant: null
  },
  step3: {
    selectedMeal: null,
    selectedServings: { value: 1, label: 1 },
    currentOrder: {},
    validMealCount: false,
    errorMessage: ""
  }
});

class DataProvider extends Component {
  state = {
    mealTypes: parseMealTypes(parseRestaurants()),
    stepOptions: defaultStepOptions()
  };

  handleChange = (selectedOption, option, currentStep) => {
    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        [currentStep]: {
          ...prevState.stepOptions[currentStep],
          [option]: selectedOption
        }
      }
    }));
  };

  addOrder = order => {
    let newOrder = { ...this.state.stepOptions.step3.currentOrder, ...order };
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder: newOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  subtractOrder = dish => {
    let currentOrder = { ...this.state.stepOptions.step3.currentOrder };
    delete currentOrder[dish];
    this.setState(
      prevState => ({
        ...prevState,
        stepOptions: {
          ...prevState.stepOptions,
          step3: {
            ...prevState.stepOptions.step3,
            currentOrder
          }
        }
      }),
      this.validateMealCount
    );
  };

  validateMealCount = () => {
    let currentOrders = this.state.stepOptions.step3.currentOrder;
    let totalMeals =
      Object.keys(currentOrders).length > 0
        ? Object.values(this.state.stepOptions.step3.currentOrder).reduce(
            (acc, val) => acc + val
          )
        : 0;

    let selectedPartyCount = this.state.stepOptions.step1.selectedPartyCount
      .value;
    let errorMessage = "";

    if (totalMeals > 10) {
      errorMessage = "Exceeded max total servings of 10.";
    }

    if (totalMeals < selectedPartyCount) {
      errorMessage =
        "Total servings must be equal to or greater than the no. of people in your group.";
    }

    this.setState(prevState => ({
      ...prevState,
      stepOptions: {
        ...prevState.stepOptions,
        step3: {
          ...prevState.stepOptions.step3,
          validMealCount: totalMeals <= 10 && totalMeals >= selectedPartyCount,
          errorMessage
        }
      }
    }));
  };

  render() {
    return (
      <DataContext.Provider
        value={{
          state: this.state,
          handleChange: (selectedOption, option, currentStep) =>
            this.handleChange(selectedOption, option, currentStep),
          addOrder: order => this.addOrder(order),
          subtractOrder: dish => this.subtractOrder(dish)
        }}
      >
        {this.props.children}
      </DataContext.Provider>
    );
  }
}

export default DataProvider;
