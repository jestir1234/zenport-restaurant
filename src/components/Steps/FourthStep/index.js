import React from "react";
import styled, { css } from "react-emotion";
import { DataContext } from "../../../context/DataProvider";
import OrderList from "../ThirdStep/OrderList";
import { MEDIA_BREAKPOINTS, COLORS } from "../../../constants";

const Container = styled("div")(props => ({
  display: "flex",
  flexDirection: "column",
  textAlign: "center",
  width: "500px",
  padding: "20px",
  border: `1px solid ${COLORS.dark}`
}));

const Row = styled("div")(props => ({
  display: "flex",
  justifyContent: "space-between",
  padding: "10px 40px 10px 40px",
  [MEDIA_BREAKPOINTS.down("md")]: {
    padding: "10px 20px 10px 20px"
  }
}));

const headerStyle = css({
  [MEDIA_BREAKPOINTS.down("md")]: {
    fontSize: "16px",
    fontWeight: "bold"
  }
});

const orderStyle = css({
  alignSelf: "center"
});

class FourthStep extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      step: "step4"
    };
  }

  render() {
    return (
      <DataContext.Consumer>
        {({ state }) => {
          let selectedMealType = state.stepOptions.step1.selectedMealType;
          let selectedPartyCount = state.stepOptions.step1.selectedPartyCount;
          let selectedRestaurant = state.stepOptions.step2.selectedRestaurant;

          if (selectedMealType) {
            selectedMealType = selectedMealType.value;
          }

          if (selectedPartyCount) {
            selectedPartyCount =
              state.stepOptions.step1.selectedPartyCount.value;
          }

          if (selectedRestaurant) {
            selectedRestaurant =
              state.stepOptions.step2.selectedRestaurant.value;
          }

          return (
            <Container>
              <h2 className={headerStyle}>Please Review Your Order</h2>
              <Row>
                <span>Meal Type</span>
                <span>{selectedMealType}</span>
              </Row>
              <Row>
                <span>No. of People</span>
                <span>{selectedPartyCount}</span>
              </Row>
              <Row>
                <span>Restaurant</span>
                <span>{selectedRestaurant}</span>
              </Row>
              <Row className={orderStyle}>
                <OrderList
                  currentOrder={state.stepOptions.step3.currentOrder}
                />
              </Row>
            </Container>
          );
        }}
      </DataContext.Consumer>
    );
  }
}

export default FourthStep;
