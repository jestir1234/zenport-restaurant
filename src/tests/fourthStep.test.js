import React from "react";
import FourthStep from "../components/Steps/FourthStep";
import renderer from "react-test-renderer";
import DataProvider from "../context/DataProvider";
import getElementWithContext from "react-test-context-provider";

const context = {
  handleChange: (e, selectedPartyCount, step) =>
    console.log("handleChangeFunction"),
  state: {
    stepOptions: {
      step1: {
        selectedPartyCount: {
          value: 4
        },
        selectedMealType: {
          value: "lunch"
        }
      },
      step2: {
        selectedRestaurant: {
          McDonalds: ["Burger", "Fries"]
        }
      },
      step3: {
        currentOrder: {
          Burger: 1,
          Fries: 1
        }
      }
    },
    mealTypes: {
      lunch: {
        restaurants: {
          McDonalds: ["Burger", "Fries"]
        }
      }
    }
  }
};

var FourthStepWithContext = getElementWithContext(
  context,
  <DataProvider>
    <FourthStep />
  </DataProvider>
);

test("Renders the <SecondStep /> component", () => {
  const component = renderer.create(FourthStepWithContext);
  let tree = component.toJSON();
  expect(tree);

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
