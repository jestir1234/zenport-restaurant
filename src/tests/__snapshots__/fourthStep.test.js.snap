// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the <SecondStep /> component 1`] = `
<div
  className="css-gzw96f"
>
  <h2
    className="css-1jmxq0i"
  >
    Please Review Your Order
  </h2>
  <div
    className="css-1gzlp2k"
  >
    <span>
      Meal Type
    </span>
    <span />
  </div>
  <div
    className="css-1gzlp2k"
  >
    <span>
      No. of People
    </span>
    <span>
      1
    </span>
  </div>
  <div
    className="css-1gzlp2k"
  >
    <span>
      Restaurant
    </span>
    <span />
  </div>
  <div
    className="css-1sraaag"
  >
    <div>
      <h3 />
      <div
        className="css-17mvcka"
      />
    </div>
  </div>
</div>
`;
